### DISCUSSION TITLE
Vector Search Technique: k-NN or ANN ?

### DISCUSSION THREAD

**Discussion by MentalGear** [oramasearch/orama#515] [Q&A] [answered]
Let me first say thank you for this superb dependency-free open source search package and great docs!

One thing the docs are missing though is an explanation how the vector search is implemented, which is important to assess accuracy and efficiency:
- Does it use k-NN or ANN as a search technique, and which algo exactly?
- Does it use a wasm library (like [this](https://yoshoku.github.io/hnswlib-node/doc/)) under the hood or is it pure js ?

**Comment by micheleriva** [comment, marked as answer]
Hi @MentalGear, vector search is implemented using cosine similarity and presumes that all the vectors have been normalized before the insertion. It's pure JS and has no external dependencies.

**Reply by MentalGear** [reply]
Thank you! May I append:
1. Does it internally store the vector's magnitude (this can lead to huge performance improvements, [example](https://github.com/andrewbloese-00/embedding-search-node#performance-gains-with-smart-vectors)) or does it re-calculate it on every compare?
2. Do vector items use an index ? Since an index is an abstraction of the actual record, this might lead to a ANN search, e.g. swapping accuracy (results may not 100% best fit / recall) with performance.

Overall, it would be great if the docs had an outline of the basic internal working of the vector search to better compare Orama with other search engines.

**Reply by micheleriva** [reply]
1. It does store the vector's magnitude for performance reasons
2. It does not use an index right now if I understood your question correctly

We'll improve our docs, thanks for pointing this out :)

**Reply by MentalGear** [reply]
Great, Orama seems like just the right match for my use case (browser usage, local database, semantic search < 100k items).
